tools/package-source.js

```diff
@@ -1712,6 +1810,64 @@ _.extend(PackageSource.prototype, {
     return _.keys(packages);
   },
 
+  // Processes the documentation provided in Package.describe. Returns an object
+  // with the following keys:
+  //   - path: full filepath to the Readme file
+  //   - excerpt: the subsection between the first and second heading of the
+  //     Readme, to be used as a longform package description.
+  //   - hash: hash of the full text of this Readme, or "" if the Readme is
+  //     blank.
+  //
+  // Returns null if the documentation is marked as null, or throws a
+  // buildmessage error if the documentation could not be read.
+  //
+  // This function reads and performs string operations on a (potentially) long
+  // file. We do not call it unless we actually need this information.
+  processReadme: function () {
+    var self = this;
+    buildmessage.assertInCapture();
+    if (! self.metadata.documentation) {
+      return null;
+    }
+
+    var ret = {};
+    ret.path =
+      files.pathJoin(self.sourceRoot, self.metadata.documentation);
+    // Read in the text of the Readme.
+    try {
+      var fullReadme = files.readFile(ret.path);
+    } catch (err) {
+      var errorMessage = "";
+      if (err.code === "ENOENT") {
+        // This is the most likely and common case, especially when we are
+        // inferring the docs as a default value.
+        errorMessage = "Documentation not found: " + self.metadata.documentation;
+      } else {
+        // This is weird, and we don't usually protect the user from errors like
+        // this, but maybe we should.
+        errorMessage =
+          "Documentation couldn't be read: " + self.metadata.documentation;
+        errorMessage += "(Error: " + err + ")";
+      }
+
+      // The user might not understand that we are automatically inferring
+      // README.md as the docs! If they want to avoid pushing anything, explain
+      // how to do that.
+      if (! self.docsExplicitlyProvided) {
+        errorMessage += "\n" +
+          "If you don't want to publish any documentation, " +
+          "please set 'documentation: null' in Package.describe";
+      }
+      buildmessage.error(errorMessage);
+      // Recover by returning null
+      return null;
+    }
+
+    ret.hash = files.fileHash(ret.path);
```

Comment:
Yeah, sorry about that. I will ping you explicitly when ready for review, I guess.